File: sdks/redlite-python/tests/test_basic.py
"""Basic tests for redlite Python SDK."""

import pytest
from redlite import Redlite, RedliteError


class TestBasic:
    """Basic functionality tests."""

    def test_open_memory(self):
        """Test opening in-memory database."""
        db = Redlite.open(":memory:")
        assert db is not None
        db.close()

    def test_context_manager(self):
        """Test context manager usage."""
        with Redlite.open(":memory:") as db:
            db.set("key", "value")
            assert db.get("key") == b"value"

    def test_version(self):
        """Test version string."""
        version = Redlite.version()
        assert version is not None
        assert "." in version


class TestStrings:
    """String command tests."""

    @pytest.fixture
    def db(self):
        with Redlite.open(":memory:") as db:
            yield db

    def test_set_get(self, db):
        """Test SET and GET."""
        db.set("key", "value")
        assert db.get("key") == b"value"

    def test_get_nonexistent(self, db):
        """Test GET on non-existent key."""
        assert db.get("nonexistent") is None

    def test_set_bytes(self, db):
        """Test SET with bytes."""
        db.set("key", b"\x00\x01\x02")
        assert db.get("key") == b"\x00\x01\x02"

    def test_incr_decr(self, db):
        """Test INCR and DECR."""
        db.set("counter", "10")
        assert db.incr("counter") == 11
        assert db.decr("counter") == 10
        assert db.incrby("counter", 5) == 15
        assert db.decrby("counter", 3) == 12

    def test_append(self, db):
        """Test APPEND."""
        db.set("key", "hello")
        length = db.append("key", " world")
        assert length == 11
        assert db.get("key") == b"hello world"

    def test_strlen(self, db):
        """Test STRLEN."""
        db.set("key", "hello")
        assert db.strlen("key") == 5
        assert db.strlen("nonexistent") == 0


class TestKeys:
    """Key command tests."""

    @pytest.fixture
    def db(self):
        with Redlite.open(":memory:") as db:
            yield db

    def test_delete(self, db):
        """Test DELETE."""
        db.set("key1", "value1")
        db.set("key2", "value2")
        assert db.delete("key1", "key2") == 2
        assert db.get("key1") is None
        assert db.get("key2") is None

    def test_exists(self, db):
        """Test EXISTS."""
        db.set("key1", "value1")
        db.set("key2", "value2")
        assert db.exists("key1") == 1
        assert db.exists("key1", "key2") == 2
        assert db.exists("key1", "nonexistent") == 1

    def test_type(self, db):
        """Test TYPE."""
        db.set("string", "value")
        db.lpush("list", "item")
        db.sadd("set", "member")

        assert db.type("string") == "string"
        assert db.type("list") == "list"
        assert db.type("set") == "set"
        assert db.type("nonexistent") == "none"

    def test_keys(self, db):
        """Test KEYS."""
        db.set("user:1", "alice")
        db.set("user:2", "bob")
        db.set("post:1", "hello")

        users = db.keys("user:*")
        assert len(users) == 2
        assert "user:1" in users
        assert "user:2" in users

    def test_dbsize(self, db):
        """Test DBSIZE."""
        assert db.dbsize() == 0
        db.set("key1", "value1")
        assert db.dbsize() == 1
        db.set("key2", "value2")
        assert db.dbsize() == 2

    def test_flushdb(self, db):
        """Test FLUSHDB."""
        db.set("key1", "value1")
        db.set("key2", "value2")
        assert db.dbsize() == 2
        db.flushdb()
        assert db.dbsize() == 0


class TestHashes:
    """Hash command tests."""

    @pytest.fixture
    def db(self):
        with Redlite.open(":memory:") as db:
            yield db

    def test_hset_hget(self, db):
        """Test HSET and HGET."""
        db.hset("hash", {"field1": "value1", "field2": "value2"})
        assert db.hget("hash", "field1") == b"value1"
        assert db.hget("hash", "field2") == b"value2"
        assert db.hget("hash", "nonexistent") is None

    def test_hset_kwargs(self, db):
        """Test HSET with kwargs."""
        db.hset("hash", name="Alice", age="30")
        assert db.hget("hash", "name") == b"Alice"
        assert db.hget("hash", "age") == b"30"

    def test_hdel(self, db):
        """Test HDEL."""
        db.hset("hash", field1="value1", field2="value2")
        assert db.hdel("hash", "field1") == 1
        assert db.hget("hash", "field1") is None
        assert db.hget("hash", "field2") == b"value2"

    def test_hexists(self, db):
        """Test HEXISTS."""
        db.hset("hash", field="value")
        assert db.hexists("hash", "field") is True
        assert db.hexists("hash", "nonexistent") is False

    def test_hlen(self, db):
        """Test HLEN."""
        db.hset("hash", field1="v1", field2="v2", field3="v3")
        assert db.hlen("hash") == 3

    def test_hkeys_hvals(self, db):
        """Test HKEYS and HVALS."""
        db.hset("hash", a="1", b="2", c="3")
        keys = db.hkeys("hash")
        vals = db.hvals("hash")
        assert len(keys) == 3
        assert len(vals) == 3
        assert set(keys) == {"a", "b", "c"}

    def test_hincrby(self, db):
        """Test HINCRBY."""
        db.hset("hash", counter="10")
        assert db.hincrby("hash", "counter", 5) == 15
        assert db.hincrby("hash", "counter", -3) == 12


class TestLists:
    """List command tests."""

    @pytest.fixture
    def db(self):
        with Redlite.open(":memory:") as db:
            yield db

    def test_lpush_rpush(self, db):
        """Test LPUSH and RPUSH."""
        db.lpush("list", "a", "b")  # list: b, a
        db.rpush("list", "c", "d")  # list: b, a, c, d
        assert db.lrange("list", 0, -1) == [b"b", b"a", b"c", b"d"]

    def test_lpop_rpop(self, db):
        """Test LPOP and RPOP."""
        db.rpush("list", "a", "b", "c")
        assert db.lpop("list") == b"a"
        assert db.rpop("list") == b"c"
        assert db.lrange("list", 0, -1) == [b"b"]

    def test_llen(self, db):
        """Test LLEN."""
        assert db.llen("list") == 0
        db.rpush("list", "a", "b", "c")
        assert db.llen("list") == 3

    def test_lindex(self, db):
        """Test LINDEX."""
        db.rpush("list", "a", "b", "c")
        assert db.lindex("list", 0) == b"a"
        assert db.lindex("list", 1) == b"b"
        assert db.lindex("list", -1) == b"c"
        assert db.lindex("list", 100) is None


class TestSets:
    """Set command tests."""

    @pytest.fixture
    def db(self):
        with Redlite.open(":memory:") as db:
            yield db

    def test_sadd_smembers(self, db):
        """Test SADD and SMEMBERS."""
        db.sadd("set", "a", "b", "c")
        members = db.smembers("set")
        assert len(members) == 3
        assert set(members) == {b"a", b"b", b"c"}

    def test_srem(self, db):
        """Test SREM."""
        db.sadd("set", "a", "b", "c")
        assert db.srem("set", "a", "b") == 2
        assert db.smembers("set") == {b"c"}

    def test_sismember(self, db):
        """Test SISMEMBER."""
        db.sadd("set", "a", "b")
        assert db.sismember("set", "a") is True
        assert db.sismember("set", "c") is False

    def test_scard(self, db):
        """Test SCARD."""
        assert db.scard("set") == 0
        db.sadd("set", "a", "b", "c")
        assert db.scard("set") == 3


class TestSortedSets:
    """Sorted set command tests."""

    @pytest.fixture
    def db(self):
        with Redlite.open(":memory:") as db:
            yield db

    def test_zadd_zscore(self, db):
        """Test ZADD and ZSCORE."""
        db.zadd("zset", {"a": 1.0, "b": 2.0, "c": 3.0})
        assert db.zscore("zset", "a") == 1.0
        assert db.zscore("zset", "b") == 2.0
        assert db.zscore("zset", "nonexistent") is None

    def test_zrem(self, db):
        """Test ZREM."""
        db.zadd("zset", a=1.0, b=2.0, c=3.0)
        assert db.zrem("zset", "a", "b") == 2
        assert db.zcard("zset") == 1

    def test_zcard(self, db):
        """Test ZCARD."""
        assert db.zcard("zset") == 0
        db.zadd("zset", a=1.0, b=2.0)
        assert db.zcard("zset") == 2

    def test_zcount(self, db):
        """Test ZCOUNT."""
        db.zadd("zset", a=1.0, b=2.0, c=3.0, d=4.0)
        assert db.zcount("zset", 2.0, 3.0) == 2

    def test_zincrby(self, db):
        """Test ZINCRBY."""
        db.zadd("zset", member=10.0)
        assert db.zincrby("zset", 5.0, "member") == 15.0
        assert db.zscore("zset", "member") == 15.0

File: sdks/redlite-zig/tests/test_basic.zig
const std = @import("std");
const redlite = @import("redlite");

const Database = redlite.Database;
const OwnedBytes = redlite.OwnedBytes;
const ZMember = redlite.ZMember;

// ============== String Commands ==============

test "SET and GET roundtrip" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key", "value", null);
    const result = try db.get("key");
    if (result) |v| {
        defer v.deinit();
        try std.testing.expectEqualSlices(u8, "value", v.data());
    } else {
        return error.TestUnexpectedResult;
    }
}

test "GET nonexistent key returns null" {
    const db = try Database.openMemory();
    defer db.close();

    const result = try db.get("nonexistent");
    try std.testing.expect(result == null);
}

test "SET with TTL" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("expiring", "value", 60);
    const ttl_result = db.ttl("expiring");
    try std.testing.expect(ttl_result > 0);
    try std.testing.expect(ttl_result <= 60);
}

test "SETEX and TTL" {
    const db = try Database.openMemory();
    defer db.close();

    try db.setex("key", 60, "value");
    const ttl_result = db.ttl("key");
    try std.testing.expect(ttl_result > 0);
    try std.testing.expect(ttl_result <= 60);
}

test "GETDEL gets and deletes" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key", "value", null);

    const result = try db.getdel("key");
    if (result) |v| {
        defer v.deinit();
        try std.testing.expectEqualSlices(u8, "value", v.data());
    } else {
        return error.TestUnexpectedResult;
    }

    // Key should be gone now
    const second = try db.get("key");
    try std.testing.expect(second == null);
}

test "APPEND" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key", "hello", null);
    const new_len = db.append("key", " world");
    try std.testing.expectEqual(@as(i64, 11), new_len);

    const result = try db.get("key");
    if (result) |v| {
        defer v.deinit();
        try std.testing.expectEqualSlices(u8, "hello world", v.data());
    }
}

test "STRLEN" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key", "hello", null);
    try std.testing.expectEqual(@as(i64, 5), db.strlen("key"));
    try std.testing.expectEqual(@as(i64, 0), db.strlen("nonexistent"));
}

test "GETRANGE" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key", "hello world", null);
    const result = db.getrange("key", 0, 4);
    defer result.deinit();
    try std.testing.expectEqualSlices(u8, "hello", result.data());
}

test "INCR and DECR" {
    const db = try Database.openMemory();
    defer db.close();

    try std.testing.expectEqual(@as(i64, 1), db.incr("counter"));
    try std.testing.expectEqual(@as(i64, 2), db.incr("counter"));
    try std.testing.expectEqual(@as(i64, 1), db.decr("counter"));
    try std.testing.expectEqual(@as(i64, 0), db.decr("counter"));
    try std.testing.expectEqual(@as(i64, -1), db.decr("counter"));
}

test "INCRBY and DECRBY" {
    const db = try Database.openMemory();
    defer db.close();

    try std.testing.expectEqual(@as(i64, 10), db.incrby("counter", 10));
    try std.testing.expectEqual(@as(i64, 15), db.incrby("counter", 5));
    try std.testing.expectEqual(@as(i64, 12), db.decrby("counter", 3));
}

// ============== Key Commands ==============

test "DEL and EXISTS" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key1", "val1", null);
    try db.set("key2", "val2", null);

    var keys = [_][:0]const u8{ "key1", "key2" };
    try std.testing.expectEqual(@as(i64, 2), db.exists(&keys));

    var del_keys = [_][:0]const u8{"key1"};
    try std.testing.expectEqual(@as(i64, 1), db.del(&del_keys));

    try std.testing.expectEqual(@as(i64, 1), db.exists(&keys));
}

test "TTL returns -2 for nonexistent key" {
    const db = try Database.openMemory();
    defer db.close();

    try std.testing.expectEqual(@as(i64, -2), db.ttl("nonexistent"));
}

test "TTL returns -1 for key without expiration" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key", "value", null);
    try std.testing.expectEqual(@as(i64, -1), db.ttl("key"));
}

test "EXPIRE and PERSIST" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key", "value", null);
    try std.testing.expect(db.expire("key", 60));

    const ttl = db.ttl("key");
    try std.testing.expect(ttl > 0);

    try std.testing.expect(db.persist("key"));
    try std.testing.expectEqual(@as(i64, -1), db.ttl("key"));
}

test "DBSIZE" {
    const db = try Database.openMemory();
    defer db.close();

    try std.testing.expectEqual(@as(i64, 0), db.dbsize());

    try db.set("key1", "val1", null);
    try db.set("key2", "val2", null);

    try std.testing.expectEqual(@as(i64, 2), db.dbsize());
}

test "FLUSHDB" {
    const db = try Database.openMemory();
    defer db.close();

    try db.set("key1", "val1", null);
    try db.set("key2", "val2", null);
    try std.testing.expectEqual(@as(i64, 2), db.dbsize());

    try db.flushdb();
    try std.testing.expectEqual(@as(i64, 0), db.dbsize());
}

// ============== Hash Commands ==============

test "HSET and HGET" {
    const db = try Database.openMemory();
    defer db.close();

    _ = db.hset("hash", "field", "value");

    const result = db.hget("hash", "field");
    if (result) |v| {
        defer v.deinit();
        try std.testing.expectEqualSlices(u8, "value", v.data());
    } else {
        return error.TestUnexpectedResult;
    }
}

test "HGET nonexistent field returns null" {
    const db = try Database.openMemory();
    defer db.close();

    const result = db.hget("hash", "nonexistent");
    try std.testing.expect(result == null);
}

test "HEXISTS" {
    const db = try Database.openMemory();
    defer db.close();

    _ = db.hset("hash", "field", "value");
    try std.testing.expect(db.hexists("hash", "field"));
    try std.testing.expect(!db.hexists("hash", "nonexistent"));
}

test "HLEN" {
    const db = try Database.openMemory();
    defer db.close();

    _ = db.hset("hash", "f1", "v1");
    _ = db.hset("hash", "f2", "v2");
    _ = db.hset("hash", "f3", "v3");

    try std.testing.expectEqual(@as(i64, 3), db.hlen("hash"));
}

test "HDEL" {
    const db = try Database.openMemory();
    defer db.close();

    _ = db.hset("hash", "f1", "v1");
    _ = db.hset("hash", "f2", "v2");

    var fields = [_][:0]const u8{"f1"};
    try std.testing.expectEqual(@as(i64, 1), db.hdel("hash", &fields));
    try std.testing.expectEqual(@as(i64, 1), db.hlen("hash"));
}

test "HINCRBY" {
    const db = try Database.openMemory();
    defer db.close();

    try std.testing.expectEqual(@as(i64, 5), db.hincrby("hash", "counter", 5));
    try std.testing.expectEqual(@as(i64, 8), db.hincrby("hash", "counter", 3));
    try std.testing.expectEqual(@as(i64, 6), db.hincrby("hash", "counter", -2));
}

// ============== List Commands ==============

test "LPUSH and RPUSH" {
    const db = try Database.openMemory();
    defer db.close();

    var values1 = [_][]const u8{"c"};
    _ = try db.lpush("list", &values1);

    var values2 = [_][]const u8{"b"};
    _ = try db.lpush("list", &values2);

    var values3 = [_][]const u8{"a"};
    _ = try db.lpush("list", &values3);

    // List should be: a, b, c
    try std.testing.expectEqual(@as(i64, 3), db.llen("list"));

    const result = db.lrange("list", 0, -1);
    defer result.deinit();
    try std.testing.expectEqual(@as(usize, 3), result.len());
    try std.testing.expectEqualSlices(u8, "a", result.get(0).?);
    try std.testing.expectEqualSlices(u8, "b", result.get(1).?);
    try std.testing.expectEqualSlices(u8, "c", result.get(2).?);
}

test "LPOP and RPOP" {
    const db = try Database.openMemory();
    defer db.close();

    var values = [_][]const u8{ "a", "b", "c" };
    _ = try db.rpush("list", &values);

    const left = db.lpop("list", 1);
    defer left.deinit();
    try std.testing.expectEqualSlices(u8, "a", left.get(0).?);

    const right = db.rpop("list", 1);
    defer right.deinit();
    try std.testing.expectEqualSlices(u8, "c", right.get(0).?);

    try std.testing.expectEqual(@as(i64, 1), db.llen("list"));
}

test "LINDEX" {
    const db = try Database.openMemory();
    defer db.close();

    var values = [_][]const u8{ "a", "b", "c" };
    _ = try db.rpush("list", &values);

    const result = db.lindex("list", 1);
    if (result) |v| {
        defer v.deinit();
        try std.testing.expectEqualSlices(u8, "b", v.data());
    } else {
        return error.TestUnexpectedResult;
    }

    // Negative index
    const last = db.lindex("list", -1);
    if (last) |v| {
        defer v.deinit();
        try std.testing.expectEqualSlices(u8, "c", v.data());
    }
}

// ============== Set Commands ==============

test "SADD and SMEMBERS" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_][]const u8{ "a", "b", "c" };
    _ = try db.sadd("set", &members);

    try std.testing.expectEqual(@as(i64, 3), db.scard("set"));

    // Adding duplicate should return 0
    var dup = [_][]const u8{"a"};
    const added = try db.sadd("set", &dup);
    try std.testing.expectEqual(@as(i64, 0), added);
}

test "SISMEMBER" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_][]const u8{ "a", "b" };
    _ = try db.sadd("set", &members);

    try std.testing.expect(db.sismember("set", "a"));
    try std.testing.expect(db.sismember("set", "b"));
    try std.testing.expect(!db.sismember("set", "c"));
}

test "SREM" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_][]const u8{ "a", "b", "c" };
    _ = try db.sadd("set", &members);

    var to_remove = [_][]const u8{"b"};
    try std.testing.expectEqual(@as(i64, 1), try db.srem("set", &to_remove));
    try std.testing.expectEqual(@as(i64, 2), db.scard("set"));
}

// ============== Sorted Set Commands ==============

test "ZADD and ZSCORE" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_]ZMember{
        .{ .score = 1.0, .member = "a" },
        .{ .score = 2.0, .member = "b" },
        .{ .score = 3.0, .member = "c" },
    };
    _ = try db.zadd("zset", &members);

    try std.testing.expectEqual(@as(i64, 3), db.zcard("zset"));
    try std.testing.expectEqual(@as(f64, 2.0), db.zscore("zset", "b").?);
}

test "ZSCORE returns null for nonexistent member" {
    const db = try Database.openMemory();
    defer db.close();

    try std.testing.expect(db.zscore("zset", "nonexistent") == null);
}

test "ZINCRBY" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_]ZMember{
        .{ .score = 1.0, .member = "a" },
    };
    _ = try db.zadd("zset", &members);

    const new_score = db.zincrby("zset", 2.5, "a");
    try std.testing.expect(@abs(new_score - 3.5) < 0.001);
}

test "ZRANGE" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_]ZMember{
        .{ .score = 3.0, .member = "c" },
        .{ .score = 1.0, .member = "a" },
        .{ .score = 2.0, .member = "b" },
    };
    _ = try db.zadd("zset", &members);

    const result = db.zrange("zset", 0, -1);
    defer result.deinit();

    try std.testing.expectEqual(@as(usize, 3), result.len());
    try std.testing.expectEqualSlices(u8, "a", result.get(0).?);
    try std.testing.expectEqualSlices(u8, "b", result.get(1).?);
    try std.testing.expectEqualSlices(u8, "c", result.get(2).?);
}

test "ZREVRANGE" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_]ZMember{
        .{ .score = 1.0, .member = "a" },
        .{ .score = 2.0, .member = "b" },
        .{ .score = 3.0, .member = "c" },
    };
    _ = try db.zadd("zset", &members);

    const result = db.zrevrange("zset", 0, -1);
    defer result.deinit();

    try std.testing.expectEqual(@as(usize, 3), result.len());
    try std.testing.expectEqualSlices(u8, "c", result.get(0).?);
    try std.testing.expectEqualSlices(u8, "b", result.get(1).?);
    try std.testing.expectEqualSlices(u8, "a", result.get(2).?);
}

test "ZCOUNT" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_]ZMember{
        .{ .score = 1.0, .member = "a" },
        .{ .score = 2.0, .member = "b" },
        .{ .score = 3.0, .member = "c" },
        .{ .score = 4.0, .member = "d" },
    };
    _ = try db.zadd("zset", &members);

    try std.testing.expectEqual(@as(i64, 2), db.zcount("zset", 2.0, 3.0));
    try std.testing.expectEqual(@as(i64, 4), db.zcount("zset", 0.0, 5.0));
}

test "ZREM" {
    const db = try Database.openMemory();
    defer db.close();

    var members = [_]ZMember{
        .{ .score = 1.0, .member = "a" },
        .{ .score = 2.0, .member = "b" },
        .{ .score = 3.0, .member = "c" },
    };
    _ = try db.zadd("zset", &members);

    var to_remove = [_][]const u8{ "a", "c" };
    try std.testing.expectEqual(@as(i64, 2), try db.zrem("zset", &to_remove));
    try std.testing.expectEqual(@as(i64, 1), db.zcard("zset"));
}
